### ISSUE TITLE
Rule-proposal: prefer-import

### ISSUE THREAD

**Issue by robcolburn** [eslint/eslint#3902]
I would ESLint to tell me to always use `import` in some project as opposed to `require`.
```js
var cat = require("cat"); // Nope
let cat = require("cat"); // Nope
const cat = require("cat"); // Nope
import cat from "cat"; // Cool
```

#### Considerations/Bike-sheds: 
* Variable `require` statements.  These can cause issue for bundlers (WebPack/Browserify), so I'm okay w/ warning (forcing user to acknowledge the un-advised pattern w/ ignore).  But, I'm also cool w/ just allowing it.
```js
const name = "thing"
require(name)

const names = ["thing1", "thing2"]
let stuff = names.map(name => require(name))
```
* `Require` only to call a function. Should we allow this case, or force an ignore rule / accept a local variable assignment?
```js
require("coolThing")()

import coolThing from "coolThing"
coolThing()
```
* `Require` to construct. Should we allow this case, or force an ignore rule / accept a local variable assignment?
```js
const cat = require("Cat")({cute: true});

import Cat from "cat"
const cat = Cat({cute: true})
```

#### Related:
* Need to avoid conflict w/ `prefer-const`
* Relates to #3478 & #3143
* Possible overlap w/ [benmosher/eslint-plugin-import](https://github.com/benmosher/eslint-plugin-import)

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by sindresorhus** [comment]
:+1:

> Variable require statements.

Until we get [`System.get()`](https://babeljs.io/docs/learn-es2015/#module-loaders) for dynamic imports, we're stuck with using `require` inline for dynamic and conditional requires, so I think there's should be some kind of an exception for this.

> Require only to call a function.

`import` can easily be used here. The only reason to use the `require('foo')()` is to save a few characters. Not worth it IMHO. Not even worth an exception.

> Require to construct.

Same as above.

**Comment by nzakas** [comment]
`require` and `import` have some subtle differences that make them not exactly equivalent, so I don't feel comfortable having this as a core rule. I'd suggest creating a custom rule for now. When Node finally implements ES6 modules, we can always revisit once we know the implications.

**Comment by sindresorhus** [comment]
@nzakas Many people use `import` today in Node through [Babel](https://github.com/babel/babel), though. I would like to enforce `import` in those cases. They're subtly different, true, but not in a way to makes a difference in most cases. I use `import` in all [my test files](https://github.com/sindresorhus/pageres/blob/771dc2a0a4a5eb866f5d0bc1dea3d73918bc4d45/test/test.js#L1-L8).

**Comment by robcolburn** [comment]
Cool, so to be sure [ES6 rules](http://eslint.org/docs/rules/#ecmascript-6) policy is based on support by Node 4.x w/o flags?

**Comment by robcolburn** [comment]
Subtle differences will likely persist, like `let/const` vs `var`, so it's up to repo maintainer to know the code style choices they prefer.  But, it's good to know ESLint's stance on rule additions.

**Comment by nzakas** [comment]
@sindresorhus that's fine, you can create a custom rule for that. 

@robcolburn nope, not at all. Each rule is evaluated on its own. The bar for core rules is high so we don't accumulate too many. We do have a doc on rule proposals here: http://eslint.org/docs/developer-guide/contributing/new-rules. 

Really, our goal is to enable developers to create their own rules. We don't want to be the gatekeepers to every rule that needs to exist.
